### PR TITLE
Add tslint rules for #3994

### PR DESCRIPTION
**Pull request by weswigham**
> - [x] One variable declaration per variable statement (see palantir/tslint#525)
> - [x] Spaces, not tabs
> - [x] No space preceding colons
> - [x] Double quotes for strings (though, depending on TSLint's behavior on this, we can be flexible.
> - [x] No explicit types for inferrable primitives (can't really use the typesystem on individual files to check for any inference)
> - [x] use namespace instead of module (see palantir/tslint#517)
> - [x] else/catch/etc. on a different line (see palantir/tslint#88)
> - [x] Space before { in interfaces, classes, enums, any sort of attached statement body, or type annotation.
> - [x] No trailing whitespace
> - [x] Forbid boolean/undefined parameters who aren't named.
> - [x] Forbid null

### CODE REVIEW

**Comment by weswigham** [comment]
Ping @DanielRosenwasser for some input. (specifically on the inference rule - is it even possible to recover that kind of info from the checker right now?)

**Comment by JKillian** [comment]
These seem like good rule concepts, any interest in contributing some/all of these to TSLint? If not, we'll probably go ahead and add at least some of them ourselves

**Comment by weswigham** [comment]
I'd probably say to go for it - which ones are you interested in?

**Comment by adidahiya** [comment]
I'm mainly interested in bringing `no-inferrable-types` into the core rules.

There seems to be interest in something like `next-line` as well, but we already have a thread going for that: https://github.com/palantir/tslint/issues/88

**Comment by JKillian** [comment]
I thought maybe no-null as well? Seems like something people might want

**Comment by adidahiya** [comment]
I'd like to wait until we see more demand for a `no-null-keyword` rule -- mainly because I'd like to create another contradictory rule that forbids implicit type casts for null checks (and _encourages_ `== null`): https://github.com/palantir/tslint/issues/625

File: scripts/tslint/noInferrableTypesRule.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                            failure = 'boolean';
+                        }
+                    break;
+                    case ts.SyntaxKind.NumberKeyword:
+                        if (e.initializer.kind === ts.SyntaxKind.NumericLiteral) {
+                            failure = 'number';
+                        }
+                    break;
+                    case ts.SyntaxKind.StringKeyword:
+                        if (e.initializer.kind === ts.SyntaxKind.StringLiteral || e.initializer.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral) {
```
Comment (DanielRosenwasser) [inline scripts/tslint/noInferrableTypesRule.ts]: *Consider* `TemplateExpression` as well

File: scripts/tslint/nextLineRule.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+        super.visitIfStatement(node);
+    }
+
+    public visitTryStatement(node: ts.TryStatement) {
+        const sourceFile = node.getSourceFile();
+        const catchClause = node.catchClause;
+
+        // "visit" try block
+        const tryKeyword = node.getChildAt(0);
```
Comment (DanielRosenwasser) [inline scripts/tslint/nextLineRule.ts]: There's a `getFirstToken()`

File: scripts/tslint/nextLineRule.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        const tryBlock = node.tryBlock;
+        const tryOpeningBrace = tryBlock.getChildAt(0);
+
+        if (this.hasOption(OPTION_CATCH) && !!catchClause) {
+            const tryClosingBrace = node.tryBlock.getChildAt(node.tryBlock.getChildCount() - 1);
```
Comment (DanielRosenwasser) [inline scripts/tslint/nextLineRule.ts]: `getLastToken()`

**Comment by weswigham** [comment]
@DanielRosenwasser feedback integrated.

**Comment by DanielRosenwasser** [comment]
:+1:

**Comment by weswigham** [comment]
@mhegazy would you like to :+1: this?

**Comment by mhegazy** [comment]
:+1:

**Comment by JKillian** [comment]
Just fyi: we're working on adding this to tslint. Feel free to leave ffedback if you have any: https://github.com/palantir/tslint/pull/690/files

I added checking for parameter declarations as well as variable declarations.

**Pull request merged**